src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,52 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> previousNode;
+
+        public Node(Node previousNode, T value) {
+            this.previousNode = previousNode;
+            this.value = value;
+        }
+    }
 
+    public void push(T value) {
+        Node newNode = new Node(null, value);
+        if (tail == null) {
+            tail = newNode;
+        } else {
+            newNode.previousNode = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     public T peek() {
-        return null;
+        isEmptyStack();
+        return tail.value;
     }
 
     public T pop() {
-        return null;
+        isEmptyStack();
+        Node temp = tail;
+        tail = tail.previousNode;
+        size--;
+        return (T) temp.value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    public void isEmptyStack() {
```

Comment:
Access modifier. Do you need public access to this method?

Reply:
actually not, private good option. done.